rfcs/records.md

```diff
@@ -0,0 +1,201 @@
+
+> TODO: How do you export a record type? `export record` would be straightforward but potentially conflicts with future export statements for functions/values.
+> Alternatively, is `export type Record = Record` too awkward?
+
+The methods defined on the record object are type checked as usual, with one exception - the implicit `self` has the type of the record. This is crucial because
+this is the one big issue we can't resolve with metatable-based OOP for tables, but it works for records because `:` is slightly more magical.
+
+When `self` is explicit, the type needs to be specified manually, e.g. these definitions are equivalent:
+
+```
+function Point:sum(): number
+    return self.x + self.y
+end
+
+function Point.sum(self: Point): number
+    return self.x + self.y
+end
+```
+
+Of course, the type checker also knows that the record type has the metatable with the inferred type of the record shape.
+
+> TODO: Does the type checker need to understand the internal structure of the shape so that type checking works across modules, or is simply modeling this
+> as a metatable sufficient?
+
+### Object modeling
+
+> TODO: records don't support implementation inheritance and why it's a good thing
+
+## Drawbacks
+
+Adding a new data type that is cross-cutting (across syntax, semantics/compiler, semantics/type checking, and runtime) results in added complexity.
+
+The rigidity of records may make some applications hesitate to adopt them; e.g. you can't simply add a new field at a random point in the program, which some would
+argue makes the language less dynamic and therefore less convenient.
+
+Not enforcing type compatibility for typed records at runtime may make it difficult for us to optimize record storage more by removing the type tags (which could
+make record objects ~2x more efficient in some cases).
+
+## Alternatives
+
+Instead of using explicit record types, we can make the VM recognize shapes of objects automatically, just like JavaScript implementations do. This requires a
+substantial amount of complicated machinery and heuristics, and likely can't be as efficient as records in the long run, but it can result in close efficiency
+without any changes to existing programs. This, however, doesn't make type safety any easier.
+
+Instead of using record types that have minimal featureset, we could implement classes that have a more feature-rich OOP semantics, with inheritance, first class
+properties, and access control. This would better map to other high level languages like TypeScript/Python, but would make the language and runtime more complicated.
+
+Instead of allowing records to have metatables, we could have separate dedicated storage for methods and come up with a new scheme for operator overloading. This
+would better map to other high level languages like C++ or C#, but would make the language less consistent.
+
```

Comment:
```suggestion

Instead of adding a new type of values, we could optimize existing tables. Tables would carry a shape and value array, as well as a hash table of values, and lookup `x.p` would be of the form
> if x's shape contains p
> then access the value array at the given offset
> otherwise use the hash table as before
which is very similar code to the current code for records. The difference semantically is that if the shape lookup fails, there is a fallback slow path rather than an error. This code can be optimized by speculating the shame in the same way as for records.

Instead of combining nominal types into records, we could provide something similar to Haskell's `newtype` declaration.
```

Reply:
This is already in the alternatives section, right?

Reply:
This is spelling it out a bit more. I don't see any discussion of alternatives to nominal types in that section.

Reply:
Sorry I meant the first paragraph/section. I'm not sure newtype is particularly relevant to records, but from the type system perspective we could easily model records as structural or nominal types - 100% of this proposal would still hold I think. I will add a section on type checking and include two alternatives.